### ISSUE TITLE
'FLIRT' and 'FNIRT' modules of FSL not present in the library

### ISSUE THREAD

**Issue by montigno** [populse/populse_mia#203]
Unless a bad installation or configuration on my part, this modules do not exist ?

Are there equivalents?

Thank you !

**Comment by servoz** [comment]
I observe the same thing on my side.
Mia provide the nipype library, as it is. It is far from being perfect (we have already pointed out [bugs to them](https://github.com/nipy/nipype), they fix them sometimes - 313 issues currently open - ...) and in some cases how the bricks work is not compatible with mia. I see that normally there should be a minimum version of [FLIRT](https://github.com/nipy/nipype/blob/028913705/nipype/interfaces/fsl/preprocess.py#L713) and [FNIRT](https://github.com/nipy/nipype/blob/028913705/nipype/interfaces/fsl/preprocess.py#L1252) in the module nipype.interfaces.fsl.preprocess. I have looked at the standard output during the automatic import of the nipype bricks when starting mia:
```
Exploring subpackages of nipype.interfaces: nipype.interfaces.fsl ...
210625-10:38:41,609 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  AR1Image
210625-10:38:41,611 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ApplyMask
210625-10:38:41,612 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ApplyTOPUP
210625-10:38:41,614 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ApplyWarp
210625-10:38:41,617 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  AvScale
210625-10:38:41,623 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brihttps://github.com/nipy/nipype/blob/028913705/nipype/interfaces/fsl/preprocess.py#L1252ck:  B0Calc
210625-10:38:41,626 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  BEDPOSTX
210625-10:38:41,632 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  BET
210625-10:38:41,636 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  BinaryMaths
210625-10:38:41,638 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ChangeDataType
210625-10:38:41,640 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Cluster
210625-10:38:41,645 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Complex
210625-10:38:41,648 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ContrastMgr
210625-10:38:41,652 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ConvertWarp
210625-10:38:41,654 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ConvertXFM
210625-10:38:41,655 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  CopyGeom
210625-10:38:41,656 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  DTIFit
210625-10:38:41,660 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  DilateImage
210625-10:38:41,662 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  DistanceMap
210625-10:38:41,663 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  EPIDeWarp
210625-10:38:41,666 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Eddy
210625-10:38:41,673 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  EddyCorrect
210625-10:38:41,674 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  EddyQuad
210625-10:38:41,677 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  EpiReg
210625-10:38:41,682 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ErodeImage
210625-10:38:41,683 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ExtractROI
210625-10:38:41,685 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FAST
210625-10:38:41,690 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FEAT
210625-10:38:41,691 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FEATModel
Detected brick:  FEATRegister
210625-10:38:41,694 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FILMGLS
210625-10:38:41,697 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FIRST
210625-10:38:41,700 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FLAMEO
210625-10:38:41,707 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
210625-10:38:41,710 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
210625-10:38:41,714 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FUGUE
210625-10:38:41,718 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FilterRegressor
210625-10:38:41,719 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  FindTheBiggest
210625-10:38:41,721 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  GLM
Detected brick:  ICA_AROMA
210625-10:38:41,729 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ImageMaths
210625-10:38:41,731 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ImageMeants
210625-10:38:41,732 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ImageStats
210625-10:38:41,734 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  InvWarp
210625-10:38:41,735 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  IsotropicSmooth
Detected brick:  L2Model
Detected brick:  Level1Design
210625-10:38:41,740 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MCFLIRT
210625-10:38:41,744 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MELODIC
210625-10:38:41,749 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MakeDyadicVectors
210625-10:38:41,750 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MaxnImage
210625-10:38:41,751 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MeanImage
210625-10:38:41,753 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MedianImage
210625-10:38:41,754 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Merge
210625-10:38:41,755 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MinImage
210625-10:38:41,757 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MotionOutliers
210625-10:38:41,758 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MultiImageMaths
Detected brick:  MultipleRegressDesign
210625-10:38:41,761 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Overlay
210625-10:38:41,763 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
/casa/home/.local/lib/python3.6/site-packages/nipype-1.6.0-py3.6.egg/nipype/interfaces/fsl/preprocess.py:2100: UserWarning: This has not been fully tested. Please report any failures.
  warn("This has not been fully tested. Please report any failures.")
Detected brick:  PRELUDE
210625-10:38:41,768 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  PercentileImage
210625-10:38:41,770 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  PlotMotionParams
210625-10:38:41,771 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  PlotTimeSeries
210625-10:38:41,774 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  PowerSpectrum
210625-10:38:41,775 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  PrepareFieldmap
210625-10:38:41,777 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ProbTrackX
210625-10:38:41,782 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ProbTrackX2
210625-10:38:41,789 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  ProjThresh
210625-10:38:41,791 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Randomise
210625-10:38:41,795 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Reorient2Std
210625-10:38:41,797 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  RobustFOV
210625-10:38:41,798 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SMM
210625-10:38:41,800 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SUSAN
210625-10:38:41,801 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SigLoss
210625-10:38:41,802 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Slice
210625-10:38:41,804 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SliceTimer
210625-10:38:41,805 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Slicer
210625-10:38:41,808 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Smooth
210625-10:38:41,809 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SmoothEstimate
210625-10:38:41,811 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SpatialFilter
210625-10:38:41,812 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Split
210625-10:38:41,814 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SwapDimensions
210625-10:38:41,815 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  TOPUP
210625-10:38:41,821 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  TemporalFilter
210625-10:38:41,823 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  Threshold
210625-10:38:41,825 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  TractSkeleton
Detected brick:  TrainingSetCreator
210625-10:38:41,828 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  UnaryMaths
210625-10:38:41,829 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  VecReg
Detected brick:  WarpPointshttps://github.com/nipy/nipype/blob/028913705/nipype/interfaces/fsl/preprocess.py#L1252
Detected brick:  WarpPointsFromStd
Detected brick:  WarpPointsToStd
210625-10:38:41,835 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  WarpUtils
210625-10:38:41,837 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  XFibres

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.aroma ...

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.base ...

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.dti ...

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.epi ...

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.fix ...

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.maths ...
210625-10:38:41,888 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MathsCommand
210625-10:38:41,891 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  MaxImage
210625-10:38:41,895 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  StdImage

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.model ...
210625-10:38:41,925 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  DualRegression

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.possum ...

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.preprocess ...

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.tests ...

Exploring subpackages of nipype.interfaces.fsl: nipype.interfaces.fsl.utils ...
210625-10:38:41,971 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SigLoss

```

So it seems that the bricks in the nipype.interfaces.fsl.preprocess are not imported at mia startup. Some things slip through the cracks ... ? Maybe this should be corrected (I haven't yet seen why the bricks are not detected).

Anyway, nipype is provided as is, contains many bricks but does not contain all the image processing softwares. We will sometimes have to code from scratch some processes. 

However, it is true that FLIRT and FNIRT are quite popular processes and should therefore be imported into mia at start-up ...

**Comment by montigno** [comment]
thank you.

So I try to create the process myself or I wait for this problem to be solved ?

**Comment by servoz** [comment]
If you feel like it, feel free to look at why FLIRT and FNIRT processes are not detected and imported into mia at the startup.
It is possible to artificially cause a new automatic search on startup by editing the process_config.yml file and setting the nipype version to nothing:
at the bottom of the file:
change for example:
```
Versions:
  capsul: 2.3.0
  mia_processes: 2.0.1-dev+a88eee2
  nipype: 1.6.0
```
by
```
Versions:
  capsul: 2.3.0
  mia_processes: 2.0.1-dev+a88eee2
  nipype:
```
I think the best solution is 
1/ try to use the nipype process when it exists,
2/ overload it in mia_processes when there is a problem (bug in nipype, way of working not compatible with mia/capsul, etc.),
3/ create from scratch the process if it is not available in nipype (and ideally push the process into mia_processes so that it is available to the whole mia's community (which we hope will be numerous as soon as possible!)

This is my point of view, but you can directly go to step 3/ if you wish.

Anyway, we'll have to see one day (I don't have the time to look at it right now) why we don't detect all the bricks in nipype !


Translated with www.DeepL.com/Translator (free version)

**Comment by montigno** [comment]
ok i will try to find the solution

**Comment by servoz** [comment]
it's strange because in nipype.interface.__init__.py we can see:
...
```
from .preprocess import (
    FAST,
    FLIRT,
    ApplyXFM,
    BET,
    MCFLIRT,
    FNIRT,
    ApplyWarp,
    SliceTimer,
    SUSAN,
    PRELUDE,
    FUGUE,
    FIRST,
)
...
```
 SliceTimer is well detected and importer in the Pipeline Manager library as:
```
210625-10:38:41,804 nipype.interface WARNING:
	 FSLOUTPUTTYPE environment variable is not set. Setting FSLOUTPUTTYPE=NIFTI
Detected brick:  SliceTimer
```
But FLIRT and FNIRT no ! As far as i can remember, there is a try / except statement when importing ? May be something go wrong at that time ? (I confess I don't remember the code in this part).

Thanks @montigno to looking at this!!!

**Comment by servoz** [comment]
@montigno , have you started to look at how to fix this issue?
Do you want to assign it to yourself (it's better to avoid two people working on it without consulting each other and therefore to avoid wasting time) or do you want to leave it to someone else?

**Comment by montigno** [comment]
Hello servoz 

I haven't looked at this problem yet, I'm learning how to make the bricks.
Maybe I'll go through the creation of bricks to have FLIRT and FNIRT.

You can start looking at this problem and I will eventually take over

**Comment by servoz** [comment]
What is your goal?
Develop alone your own pipeline?
I thought you wanted to work in the populse project (there is still a lot of work to do on mia right now)?

**Comment by montigno** [comment]
Indeed, if there is still a lot of work to be done on MIA, I must focus on its development as a priority.

Ok, I'm starting to look at this problem with FLIRT and FNIRT.

**Comment by servoz** [comment]
Sorry, I don't want to stop you from developing your pipelines ...
That's the goal we all want to achieve as soon as possible!!!
 FYI, it's been a year since I've worked on my own bricks, since there are still so much things to do in populse and mia particularly ... Among other things, yes, there are still many tickets that should be closed, for the good of all.

Don't feel forced to work on this ticket, but it's true that if you do I'll be very grateful :-)))
Also, it is related to the pipeline you need... If the FLIRT and FNIRT bricks are available in nipype, you might as well try them first and see what happens. Sometimes you only need to change a few things to make it work. This avoids creating the brick from scratch, which requires more work.

About the creation of bricks from scratch. You can do it by yourself if you want, but I advise you to talk about it first in order not to waste time. Indeed there are several ways to do it and I think the best would be to go through a capsul process ... I and @denisri can certainly save your time by explaining what is best to do.

In any case, thank you very much if you fix this issue (it will be one less!!! and mia will be all the better!!!).
Don't hesitate to communicate, with several people we are stronger than alone !!!

**Comment by montigno** [comment]
No problem I continue to watch from time to time the construction of the bricks but the problems of MIA are priority.

I have a question : in the casa distro container, where are the python libraries installed?
I typed the command `python3 -c 'import sysconfig; print(sysconfig.get_paths()["purelib"])'` and it displays `/usr/lib/python3.6/site-packages` but the 'site-packages' folder doesn't exist !  (the virtual environment is actived)

**Comment by servoz** [comment]
`'sysconfig.get_paths()["purelib"]'`, returns an installation path from the installation scheme in the station and for global site packs.

So this means that in the case of a global installation (different from the per user site-packages case), the installation would actually take place in  '/usr/lib/python3.6/site-packages'.

There are no site-packages because the installation according to the global shema has never been done, I think. On the other hand, on my station there is a dist-packages directory :
```
(opensource-master-5.0)econdami@irmage-pteric:/usr/local/lib/python3.6/dist-packages$ pip3 show pyqt5
Name: PyQt5
Version: 5.12.1
Summary: None
Home-page: None
Author: None
Author-email: None
License: None
Location: /usr/local/lib/python3.6/dist-packages
Requires: 

```
Generally I use the per user site-packages scheme (--user). This avoids mixing the libraries of the system/distibution and those of the users. Some people say that it is criminal to use su or sudo to do a pip installation (it will go into the system's package site). With the per user site-packages scheme, no need of su/sudo and the package go on your site-package (on the container, in /casa/home/.local/lib/python3.6/site-packages).

I think that for the casa distro container especially, maybe @denisri  can tell you more, because he knows it better than I do.

**Comment by montigno** [comment]
Indeed, I have   `/usr/local/lib/python3.6/dist-packages/`, that's surely it.


Otherwise, when I update the nipype library (with the 'File / Package library manager' menu), I get some interesting lines in the terminal : 

<p></p>

	Error during installation of the "nipype.interfaces.fsl" module ...!
	Traceback:
  	File "/home/omontigon/Apps/casa_distro/populse_mia/python/populse_mia/user_interface/pipeline_manager/process_library.py", line 1516, in add_package
    	'%s.%s' % (module_name, v.__name__))
  	File "/casa/host/src/capsul/master/capsul/study_config/process_instance.py", line 142, in get_process_instance
    	**kwargs)
  	File "/casa/host/src/capsul/master/capsul/study_config/process_instance.py", line 343, in _get_process_instance
    	result = nipype_factory(module_object())
  	File "/casa/host/src/capsul/master/capsul/process/nipype_process.py", line 328, in nipype_factory
    	getattr(nipype_instance.inputs, trait_name))
	AttributeError: 'FLIRTInputSpec' object has no attribute 'nipype_save_log'


	Error during installation of the "nipype.interfaces.fsl" module ...!
	Traceback:
  	File "/home/omontigon/Apps/casa_distro/populse_mia/python/populse_mia/user_interface/pipeline_manager/process_library.py", line 1516, in add_package
    	'%s.%s' % (module_name, v.__name__))
  	File "/casa/host/src/capsul/master/capsul/study_config/process_instance.py", line 142, in get_process_instance
    	**kwargs)
  	File "/casa/host/src/capsul/master/capsul/study_config/process_instance.py", line 343, in _get_process_instance
    	result = nipype_factory(module_object())
  	File "/casa/host/src/capsul/master/capsul/process/nipype_process.py", line 328, in nipype_factory
    	getattr(nipype_instance.inputs, trait_name))
	AttributeError: 'FNIRTInputSpec' object has no attribute 'nipype_log_file'



The error would come from the 'nipype_process.py' file in Capsul. Denis can maybe confirm?

Thank you !

**Comment by servoz** [comment]
In the capsul.process.nipype_process.nipype_factory(), we try to initialise the process_instance values with the initial values of the nipype interface and [we currently only catch the TraitError exception](https://github.com/populse/capsul/blob/master/capsul/process/nipype_process.py#L329).
If we focus only on the FLIRT and FNIRT fsl bricks, the exception thrown is of AttributeError type for the `save_log` and `log_file` initial parameters of the nipype interface, respectively.
I haven't taken the time to look at why these parameters were not created in the process_instance (maybe because of the nipype part?) but a first patch could be to catch the AttributeError exception as well.
The first test seems to indicate that in this case, these 2 processes/bricks are now available. I haven't checked if these 2 bricks/processes work after this patch.
I think to properly fix this ticket, we would also need to understand why these parameters are not created before for the process_instance object.
I also don't know if this small change can have any side effects?
I made a [PR](https://github.com/populse/capsul/pull/181) in capsul with this small change, so we can test the possible side effects before a possible merge.
Beyond the FLIRT and FNIRT bricks, this ticket indicates that it might be necessary to check the automatic provision step of the bricks in the manager pipeline.

**Comment by servoz** [comment]
I just pushed a [cleaner fix on the Fix4Populse#203 capsul branch](https://github.com/populse/capsul/commit/0a1a181891ee8d4dc7e1cba9d59ec1ecbac9d667) for this ticket (a [PR is pending with this branch in capsul](https://github.com/populse/capsul/pull/181)).

In a nutshell:
-  if a trait name exists in the process_instance and in the nipype_instance (ex save_log is a function of the process_instance and save_log is also an input parameter of the nipype_instance), the nipype_factory prepends the trait name with `nipype_` in process_instance. The problem was that the getters and setters fo the nipype_instance then used the `nypipe_`xxx name which raised an AttributeError exception. The latest change to the RP more cleanly fixes this issue (and we're back to only TraitError exception catch).

I have done some tests which seem to indicate that there is no side effect. I have only tested the FLIRT process which seems to work fine.

I think this ticket can now be closed.
@montigno, can you test on your side that everything works well (waiting for the PR to be merged into capsul, you can use the capsul Fix4Populse#203 branch) ?
It is possible to artificially cause a new automatic search at startup by editing the process_config.yml file and setting the nipype version to nothing:
at the bottom of the file:
change for example:

```
Versions:
  capsul: 2.3.0
  mia_processes: 2.0.1-dev+a88eee2
  nipype: 1.6.0
```

by

```
Versions:
  capsul: 2.3.0
  mia_processes: 2.0.1-dev+a88eee2
  nipype:

```

 Don't hesitate to reopen this ticket if you have a problem!
